### ISSUE TITLE
Crashing periodically in Tectonic 1.6.2

### ISSUE THREAD

**Issue by robszumski** [coreos/container-linux-update-operator#75]
This doesn't seem normal:

```
$ kubectl get pods
NAME                                              READY     STATUS    RESTARTS   AGE
container-linux-update-agent-ds-fs5c6             1/1       Running   195        5d
container-linux-update-agent-ds-j0bz3             1/1       Running   197        5d
container-linux-update-agent-ds-shcw0             1/1       Running   205        5d
container-linux-update-agent-ds-sj9ds             1/1       Running   190        5d
container-linux-update-agent-ds-vv8w2             1/1       Running   195        5d
```

Logs from a running instance look ok:

```
$ kubectl logs container-linux-update-agent-ds-fs5c6
I0518 18:08:12.158135       1 main.go:42] /bin/update-agent running
I0518 18:08:12.158212       1 agent.go:117] Setting info labels
I0518 18:08:12.195741       1 agent.go:123] Marking node as schedulable
I0518 18:08:12.208552       1 agent.go:134] Setting annotations map[string]string{"container-linux-update.v1.coreos.com/reboot-in-progress":"false", "container-linux-update.v1.coreos.com/reboot-needed":"false"}
I0518 18:08:12.225767       1 agent.go:148] Waiting for ok-to-reboot from controller...
I0518 18:08:12.225865       1 agent.go:205] Beginning to watch update_engine status
I0518 18:08:12.226547       1 agent.go:68] Updating status
```

Logs from a previous run: 

```
$ kubectl logs container-linux-update-agent-ds-fs5c6 -p
I0518 17:28:51.269685       1 main.go:42] /bin/update-agent running
I0518 17:28:51.269748       1 agent.go:117] Setting info labels
I0518 17:28:51.319913       1 agent.go:123] Marking node as schedulable
I0518 17:28:51.336287       1 agent.go:134] Setting annotations map[string]string{"container-linux-update.v1.coreos.com/reboot-in-progress":"false", "container-linux-update.v1.coreos.com/reboot-needed":"false"}
I0518 17:28:51.357811       1 agent.go:148] Waiting for ok-to-reboot from controller...
I0518 17:28:51.357976       1 agent.go:205] Beginning to watch update_engine status
I0518 17:28:51.358551       1 agent.go:68] Updating status
I0518 17:40:40.674321       1 agent.go:68] Updating status
I0518 17:40:50.693834       1 agent.go:68] Updating status
F0518 18:08:11.590943       1 main.go:45] Error while running /bin/update-agent: waiting for annotation "container-linux-update.v1.coreos.com/reboot-ok" failed: watch closed before Until timeout
```

Doesn't seem like that last log should cause it to crash?

**Comment by crawford** [comment]
If the agent is crashing, it's likely something is wrong with the controller. Is the controller still running properly?

**Comment by robszumski** [comment]
Appears to be running and there is no restart count increasing:

```
$ kubectl logs container-linux-update-operator-886082281-hvrlj
I0512 19:08:11.751763       1 main.go:46] /bin/update-operator running
I0512 19:08:11.922983       1 leaderelection.go:179] attempting to acquire leader lease...
I0512 19:08:11.948075       1 leaderelection.go:189] successfully acquired lease tectonic-system/container-linux-update-operator-lock
```

**Comment by robszumski** [comment]
Bump on this

**Comment by sym3tri** [comment]
Any progress on this?

**Comment by crawford** [comment]
Looks like we just need to check to see if `err == watch.errWatchClosed`-- oh wait, it's not exported.

**Comment by sym3tri** [comment]
@crawford 
So CL updates just don't work on Tectonic until this is fixed? Can be bump the priority on this?

**Comment by crawford** [comment]
@sym3tri no, updates work fine. When the watch times out, this crashes and restarts. It picks up where it left off and continues working. In my mind, this is a very low priority.

**Comment by metral** [comment]
I'm hitting this issue as well on Azure

**Comment by colhom** [comment]
@crawford 

> In my mind, this is a very low priority.

Pod restarts are often used as a proxy variable for health of the cluster- having them occur regularly in an otherwise healthy cluster is seen as cause for concern. Especially with control plane components, this indicates to the user that something is wrong.

Aside from error checking `watch.Until`, we also can't have update-agent exiting when API watches conclude in general. https://github.com/coreos/container-linux-update-operator/issues/79

**Comment by crawford** [comment]
Now that https://github.com/kubernetes/kubernetes/pull/46449 has been merged, we can go in and fix up this logic.

**Comment by sym3tri** [comment]
Would be great if we could get this into the next Tectonic release. Code freeze for that is June 16th.

**Comment by euank** [comment]
I checked the most recent errors on my 4-node cluster, which confirms everything above as well.

4x `main.go:45] Error while running /bin/update-agent: waiting for annotation "container-linux-update.v1.coreos.com/reboot-ok" failed: watch closed before Until timeout`

I suspect that ELB's behaviour of semi-frequently dropping long-lived connections is related to why this is so frequent.

Now that the error's exported, I'll give fixing this a shot.

**Comment by robszumski** [comment]
For anyone following along, this is still happening in `1.6.4-tectonic.1`

**Comment by dghubble** [comment]
@crawford pulling in that apimachinery fix conflicts with client-go v3.0.0-beta.0. We'll need to update to something closer to master or wait for a tagged release.

**Comment by euank** [comment]
I made a more simplistic change which retries on all errors rather than distinguishing this one expected error. I think that should be fine to consider this fixed.

**Comment by dghubble** [comment]
CLUO v0.2.1 is included for the next Tectonic release (1.6.6-tectonic.1)
